src/service.rs

```diff
@@ -299,6 +312,8 @@ impl Service {
                 let cmd_sender3 = cmd_sender2.clone();
 
                 let _ = cmd_sender2.send(Box::new(move |state: &mut State| {
+                    state.listening_ports.remove(&port);
```

Comment:
Sure. Every time the service is about to accept a new connection, the port the acceptor listens on is stored in this hashset (see line 304). Then once something is accepted, the port is removed. So the hashset keeps tracks of the ports the acceptors are currently holding while they are blocked waiting for incoming connection. This hashset is then used in the `stop` method (lines 227 to 234) where dummy connects are made to these ports, to "wake up" the blocked acceptors, so they can be destroyed.